### PR TITLE
Add printing out debug logs using external output

### PR DESCRIPTION
**Pull request by daeyeon**
IoT.js-DCO-1.0-Signed-off-by: Daeyeon Jeong daeyeon.jeong@samsung.com

### CODE REVIEW

**Comment by glistening** [comment]
@daeyeon What about adding your implementation for debug log in existing  `src/iotjs_debuglog.c` and call the function in console module?

File: src/iotjs_env.c [hunk 1]
```diff
@@ -152,6 +153,15 @@ uv_loop_t* iotjs_environment_loop(const iotjs_environment_t* env) {
   return _this->loop;
 }
 
+void iotjs_environment_set_dbgout(iotjs_environment_t* env, dbgout_t dbgout) {
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_environment_t, env);
```
Comment (glistening) [inline src/iotjs_env.c]: @daeyeon Currently we are removing `IOTJS_VALIDATED_STRUCT_XXX`. We can remove it.

File: src/iotjs_env.c [hunk 2]
```diff
@@ -152,6 +153,15 @@ uv_loop_t* iotjs_environment_loop(const iotjs_environment_t* env) {
...
+  _this->dbgout = dbgout;
+}
+
+dbgout_t iotjs_environment_dbgout(const iotjs_environment_t* env) {
+  const IOTJS_VALIDATED_STRUCT_METHOD(iotjs_environment_t, env);
```
Comment (glistening) [inline src/iotjs_env.c]: @daeyeon Ditto

File: src/iotjs_env.c (hunk 1 above)
Comment (glistening) [reply src/iotjs_env.c]: @daeyeon I don't want to mix 'validated struct' removal patch with your PR. So I pushed #1432.

**Comment by daeyeon** [comment]
@glistening `iotjs_debuglog.c` is updated.

File: src/iotjs_env.c (hunk 1 above)
Comment (daeyeon) [reply src/iotjs_env.c]: I agree. It should be handled in another pr.

File: src/iotjs_debuglog.h [hunk 3]
```diff
@@ -36,6 +37,8 @@ extern const char* iotjs_debug_prefix[4];
       fprintf(iotjs_log_stream, __VA_ARGS__);                       \
       fprintf(iotjs_log_stream, "\n");                              \
       fflush(iotjs_log_stream);                                     \
+      if(iotjs_debug_extern_out)                                    \
```
Comment (glistening) [inline src/iotjs_debuglog.h]: @daeyeon Please add a space between if and opening parenthesis.

File: src/modules/iotjs_module_console.c [hunk 4]
```diff
@@ -15,6 +15,10 @@
 
 #include "iotjs_def.h"
 
+#define DBG_BUF_SIZE 1024
```
Comment (glistening) [inline src/modules/iotjs_module_console.c]: @daeyeon Too big.

File: src/iotjs_debuglog.h (hunk 3 above)
Comment (daeyeon) [reply src/iotjs_debuglog.h]: fixed.

File: src/modules/iotjs_module_console.c [hunk 5]
```diff
@@ -25,13 +29,26 @@ static jerry_value_t Print(const jerry_value_t* jargv,
...
-    } else {
-      fprintf(out_fd, "\\u0000");
+  iotjs_environment_t* env = iotjs_environment_get();
+  dbgout_t dbgout = iotjs_environment_dbgout(env);
+
+  if (dbgout) {
+    char tmpbuf[DBG_BUF_SIZE];
+    int level = (out_fd == stdout) ? DBG_LV_NORMAL : DBG_LV_ERR;
+
+    strncpy(tmpbuf, str, DBG_BUF_SIZE - 1);
```
Comment (glistening) [inline src/modules/iotjs_module_console.c]: @daeyeon I think we don't need `strncpy`. Therefore we can get rid of `tmpbuf`.

File: src/iotjs_debuglog.h [hunk 6]
```diff
@@ -36,6 +37,8 @@ extern const char* iotjs_debug_prefix[4];
       fprintf(iotjs_log_stream, __VA_ARGS__);                       \
       fprintf(iotjs_log_stream, "\n");                              \
       fflush(iotjs_log_stream);                                     \
+      if (iotjs_debug_extern_out)                                   \
+        iotjs_debug_extern_out(lvl, __VA_ARGS__);                   \
```
Comment (glistening) [inline src/iotjs_debuglog.h]: @daeyeon What is the difference between `iotjs_log_stream` and `iotjs_debug_extern_out`? 

First, I cannot understand why do we need two separate logging systems.

Second, implementation are scattered in inconsistent way. `iotjs_log_stream` is defined as global variable in iotjs_debuglog.c and `iotjs_debug_extern_out` is defined in iotjs_env as a member. It makes me understand and follow up the code.

**Pull request closed without merging**